py/obj.h: Add m_new_obj_with_type macro.

jimmo opened the pull request:
This is to replace the following:

```
mp_foo_obj_t *self = m_new_obj(mp_foo_obj_t);
self->base.type = &mp_type_foo;
```

with:

```
mp_foo_obj_t *self = m_new_obj_with_type(mp_foo_obj_t, &mp_type_foo);
```

Calling the function is less code than inlining setting the type everywhere, adds up to 100 bytes on PYBV11. Small, but every byte counts, right? :p

It also helps to avoid an easy mistake of forgetting to set the type.

Perf impact is negligible, possibly even improved? I specifically excluded `mp_obj_new_float` from using this (otherwise there was about a 1% decrease across the fp-heavy tests).

```
$ ./run-perfbench.py -s ~/obj-type-pybv11-baseline ~/obj-type-pybv11-helper-float 
diff of scores (higher is better)
N=100 M=100                /home/jimmo/obj-type-pybv11-baseline -> /home/jimmo/obj-type-pybv11-helper-float         diff      diff% (error%)
bm_chaos.py                    358.07 ->     366.90 :      +8.83 =  +2.466% (+/-0.01%)
bm_fannkuch.py                  76.62 ->      77.21 :      +0.59 =  +0.770% (+/-0.01%)
bm_fft.py                     2542.13 ->    2527.94 :     -14.19 =  -0.558% (+/-0.00%)
bm_float.py                   5916.09 ->    6066.47 :    +150.38 =  +2.542% (+/-0.02%)
bm_hexiom.py                    43.71 ->      43.80 :      +0.09 =  +0.206% (+/-0.00%)
bm_nqueens.py                 4398.49 ->    4363.18 :     -35.31 =  -0.803% (+/-0.00%)
bm_pidigits.py                 640.04 ->     637.61 :      -2.43 =  -0.380% (+/-0.22%)
misc_aes.py                    412.97 ->     421.05 :      +8.08 =  +1.957% (+/-0.01%)
misc_mandel.py                3520.42 ->    3521.73 :      +1.31 =  +0.037% (+/-0.00%)
misc_pystone.py               2275.25 ->    2302.60 :     +27.35 =  +1.202% (+/-0.00%)
misc_raytrace.py               376.95 ->     386.30 :      +9.35 =  +2.480% (+/-0.01%)
```

jimmo committed:
py/obj.h: Add mp_obj_malloc macro.

This is to replace the following:

  mp_foo_obj_t *self = m_new_obj(mp_foo_obj_t);
  self->base.type = &mp_type_foo;

with:

  mp_foo_obj_t *self = mp_obj_malloc(mp_foo_obj_t, &mp_type_foo);

Calling the function is less code than inlining setting the type
everywhere, adds up to ~100 bytes on PYBV11.

It also helps to avoid an easy mistake of forgetting to set the type.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

robert-hh commented:
On some port & boards indeed every byte counts. b.t.w. In the mimxrt port seemed to have missed that code in machine_pin.c and machine_i2s.c

dpgeorge commented on py/obj.h:
```diff
@@ -732,6 +732,12 @@ extern const struct _mp_obj_exception_t mp_const_GeneratorExit_obj;
+// Implementing this as a call rather than inline saves 8 bytes per usage.
+#define m_new_obj_with_type(struct_type, obj_type) ((struct_type *)_m_new_obj_with_type(sizeof(struct_type), obj_type))
+#define m_new_obj_var_with_type(struct_type, var_type, var_num, obj_type) ((struct_type *)_m_new_obj_with_type(sizeof(struct_type) + sizeof(var_type) * (var_num), obj_type))
+void *_m_new_obj_with_type(size_t num_bytes, const mp_obj_type_t *type);
```
Identifiers cannot start with `_`, that's reserved for the compiler/C runtime to use.

Also, do these belong in `misc.h` alongside the `m_new_obj`?

jimmo replied:
Updated -- renamed to `mp_obj_malloc{,_var}` and the helper (in obj.c) is `mp_obj_malloc_helper`.

dpgeorge commented on py/obj.c:
```diff
@@ -602,3 +602,9 @@ mp_obj_t mp_generic_unary_op(mp_unary_op_t op, mp_obj_t o_in) {
 }
+
+void *_m_new_obj_with_type(size_t num_bytes, const mp_obj_type_t *type) {
+    mp_obj_base_t *base = (mp_obj_base_t *)m_malloc(num_bytes);
```
Maybe `m_malloc()` can take the type (or have a new version of `m_malloc()`), then it would eliminate a C call altogether?

dpgeorge replied:
`m_malloc()` is pretty small, wouldn't cost much to duplicate it to `m_malloc_typed(...)`.

jimmo replied:
There was a very small performance gain for this but it cost almost all of the byte saving.

dpgeorge commented on extmod/machine_i2c.c:
```diff
@@ -668,8 +668,7 @@ STATIC void mp_machine_soft_i2c_init(mp_obj_base_t *self_in, size_t n_args, cons
     // create new soft I2C object
-    machine_i2c_obj_t *self = m_new_obj(machine_i2c_obj_t);
-    self->base.type = &mp_machine_soft_i2c_type;
+    machine_i2c_obj_t *self = m_new_obj_with_type(machine_i2c_obj_t,  &mp_machine_soft_i2c_type);
```
What's up with the double space?  And why didn't code formatting pick that up?

jimmo committed:
top: Use mp_obj_malloc everywhere.

This replaces occurences of
  foo_t *foo = m_new_obj(foo_t);
  foo->base.type = &foo_type;
with
  foo_t *foo = mp_obj_malloc(foo_t, &foo_type);

Excludes any places where base is a sub-field or when new0/memset is used.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

dpgeorge commented:
please rebase on latest master to pick up new renesas-ra port, which has 7 uses of `m_new_obj()`

jimmo commented:
Done. Not all 7 were suitable (two in timer.c immediately memset, and in mpthreadport.c it isn't actually an object).

dpgeorge commented:
The first commit includes changes to `py/objfloat.c` and `py/objset.c` which I think should be in the second commit (and the set change is wrong, uses an old version of the marco).

In the second commit some cc3200 HAL files are unnecessarily changed (maybe line endings?).

jimmo committed:
py/objfloat.c: Explain why mp_obj_malloc isn't used.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo commented:
Done.

dpgeorge commented:
Thanks for fixing.  Merged in 709e8328d9812a2e02da6fba65a03f9a873d60a2 through 0e7bfc88c6ac6b5d64240f91183a3cfe2ab67ade
